eth/tracers/native/call_flat.go

```diff
@@ -0,0 +1,447 @@
+}
+
+// newFlatCallTracer returns a new mux tracer.
+func newFlatCallTracer(ctx *tracers.Context, cfg json.RawMessage) (tracers.Tracer, error) {
+	var config flatCallTracerConfig
+	if cfg != nil {
+		if err := json.Unmarshal(cfg, &config); err != nil {
+			return nil, err
+		}
+	}
+
+	tracer, err := tracers.New("callTracer", ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &flatCallTracer{tracer: tracer, ctx: ctx, config: config}, nil
+}
+
+// CaptureStart implements the EVMLogger interface to initialize the tracing operation.
+func (t *flatCallTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	t.tracer.CaptureStart(env, from, to, create, input, gas, value)
+}
+
+// CaptureEnd is called after the call finishes to finalize the tracing.
+func (t *flatCallTracer) CaptureEnd(output []byte, gasUsed uint64, elapsed time.Duration, err error) {
+	t.tracer.CaptureEnd(output, gasUsed, elapsed, err)
+}
+
+// CaptureState implements the EVMLogger interface to trace a single step of VM execution.
+func (t *flatCallTracer) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
+	t.tracer.CaptureState(pc, op, gas, cost, scope, rData, depth, err)
+}
+
+// CaptureFault implements the EVMLogger interface to trace an execution fault.
+func (t *flatCallTracer) CaptureFault(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, depth int, err error) {
+	t.tracer.CaptureFault(pc, op, gas, cost, scope, depth, err)
+}
+
+// CaptureEnter is called when EVM enters a new scope (via call, create or selfdestruct).
+func (t *flatCallTracer) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	t.tracer.CaptureEnter(typ, from, to, input, gas, value)
+}
+
+// CaptureExit is called when EVM exits a scope, even if the scope didn't
+// execute any code.
+func (t *flatCallTracer) CaptureExit(output []byte, gasUsed uint64, err error) {
+	t.tracer.CaptureExit(output, gasUsed, err)
+}
+
+func (t *flatCallTracer) CaptureTxStart(gasLimit uint64) {
+	t.tracer.CaptureTxStart(gasLimit)
+}
+
+func (t *flatCallTracer) CaptureTxEnd(restGas uint64) {
+	t.tracer.CaptureTxEnd(restGas)
+}
+
+// GetResult returns an empty json object.
+func (t *flatCallTracer) GetResult() (json.RawMessage, error) {
+	traceResultJson, err := t.tracer.GetResult()
+	if err != nil {
+		return nil, err
+	}
+
+	traceResult := new(callFrame)
+	err = json.Unmarshal(traceResultJson, &traceResult)
+	if err != nil {
+		return nil, err
+	}
+
+	// not so nice way to read the TypeString from the json
+	traceResultMarshaled := new(callFrameMarshaling)
+	err = json.Unmarshal(traceResultJson, &traceResultMarshaled)
+	if err != nil {
+		return nil, err
+	}
+
+	traceResult.Type = vm.StringToOp(traceResultMarshaled.TypeString)
```

Comment:
Here we do this ugly hack to read, the **type** from the json.RawMessage passed from **callTracer**, where **type** is a string.
The thing is that if we don't find a better way to do that, this way has to be extended to do it recursively for the inners calls as well.

Reply:
Resolved by https://github.com/etclabscore/ethereum.go-ethereum/pull/14/commits/271a8b897bf1650b5a51f423133e34aee1bc0d22